File: problems/CSES/Dynamic_Programming/array_description/array_description.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    MOD = 10 ** 9 + 7

    n, m = map(int, input().split())
    arr = list(map(int, input().split()))

    prev = [0] * (m + 2)

    if arr[0] == 0:
        for value in range(1, m + 1):
            prev[value] = 1
    else:
        prev[arr[0]] = 1

    for i in range(1, n):
        curr = [0] * (m + 2)

        if arr[i] == 0:
            for value in range(1, m + 1):
                curr[value] = (prev[value - 1] + prev[value] + prev[value + 1]) % MOD
        else:
            value = arr[i]
            curr[value] = (prev[value - 1] + prev[value] + prev[value + 1]) % MOD

        prev = curr

    print(sum(prev) % MOD)

main()

File: problems/CSES/Dynamic_Programming/array_description/array_description.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr int MOD = 1'000'000'007;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<int> arr(n);

    for (int& value : arr) {
        cin >> value;
    }

    vector<int> prev(m + 2, 0);

    if (arr[0] == 0) {
        for (int value = 1; value <= m; ++value) {
            prev[value] = 1;
        }
    } else {
        prev[arr[0]] = 1;
    }

    for (int i = 1; i < n; ++i) {
        vector<int> curr(m + 2, 0);

        if (arr[i] == 0) {
            for (int value = 1; value <= m; ++value) {
                curr[value] = (
                    1LL * prev[value - 1] +
                    prev[value] +
                    prev[value + 1]
                ) % MOD;
            }
        } else {
            int value = arr[i];

            curr[value] = (
                1LL * prev[value - 1] +
                prev[value] +
                prev[value + 1]
            ) % MOD;
        }

        prev = move(curr);
    }

    int sum = 0;

    for (int value = 1; value <= m; ++value) {
        sum += prev[value];

        if (sum >= MOD) {
            sum -= MOD;
        }
    }

    cout << sum << '\n';

    return 0;
}
